cudaPackages: build redists from manifests and add CUDA 13

ConnorBaker opened the pull request:
> [!IMPORTANT]
>
> The addition of CUDA 13 **does not mean** packages will suddenly work with CUDA 13. Expect breakages.

Largely based on work done in https://github.com/ConnorBaker/cuda-packages.

Big changes:

- CUDA 13 support
- Introduced manifests for many redistributables in `_cuda.manifests`
  - cublasmp
  - cuda
  - cudnn
  - cudss
  - cuquantum
  - cusolvermp
  - cusparselt
  - cutensor
  - nppplus
  - nvcomp
  - nvjpeg2000
  - nvpl
  - nvshmem
  - nvtiff
  - tensorrt (NOTE: TensorRT doesn't have manifests but there is documentation and a helper script for generating them)
- Turned `_cuda.fixups` into full package expressions which exist as files, in-tree
- Introduction of additional NVIDIA licenses in `_cuda.lib.licenses`
- Updated `_cuda.lib.allowUnfreeCudaPredicate` to use `_cuda.lib.licenses`
- Updated CUDA docs showing how to use `_cuda.lib.allowUnfreeCudaPredicate` by having `config` take an attribute set containing `pkgs` as an argument 
- ~~Enabling `config.cudaSupport` automatically adds NVIDIA's licenses to `config.allowlistedLicenses`, avoiding the need to set `config.allowUnfree`~~
  - This has been reverted; such a change is contested (see discussion in https://github.com/NixOS/nixpkgs/pull/437723#discussion_r2361011957) and not strictly required for this PR
- Module system evaluation as part of CUDA package set creation has been removed
- Consistent attribute sets across CUDA package set versions as a result of using package expressions for redistributables
- Removal of "feature" manifests (this information is now encoded directly in redistributable package expressions)
- Introduced `buildRedist` helper function
- Removed usage of `overrideAttrs`
- Introduced `tests.redists-unpacked` and `tests.redists-installed` to verify downloading all sources and unpacking and installing all redistributables work as intended
- `_cuda.lib.{_mkMetaBroken,_mkMetaBadPlatforms}` have been updated to use `builtins.traceVerbose` to avoid polluting output when doing whole-tree evaluations (e.g., during `nixpkgs-review` runs) and now take one argument instead of two
- Introduced `cudaPackages.tests.{cmake,cudnn-frontend,onnx-tensorrt}`

In the long run, I absolutely think that a database-backed approach like https://github.com/NixOS/nixpkgs/pull/406740 is the right way to handle package creation and generation. Unfortunately, it's not quite ready. In the interim, an expression per-redistributable package enables easily finding the relevant file in-tree (no flattening of deeply nested JSON), consistent package set members across versions (instead of adding attributes on the fly for supported systems), and avoiding runtime generation of package expressions.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

As always with `nixpkgs-review`, part of the problem is the unfree packages aren't built or cached and so they're typically rebuilt. Using `allowUnfree` enables many more packages to be built than just the CUDA packages.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ConnorBaker committed:
onnxruntime: properly condition ncclSupport

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
python3Packages.torch: fix condition and usage of static NCCL

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker commented:
What do you mean by "system-wide"?

cpcloud commented:
I've set `config.cudaSupport` to true in my flake's nixpkgs instance.

ConnorBaker commented:
Do you have an example flake you can post? Also, have you set `config.allowUnfree` to true? See https://nixos.org/manual/nixpkgs/stable/#cuda-configuring-nixpkgs-for-cuda for an example (though it uses the more specific `allowUnfreePredicate`).

With this PR, I can do

```
NIXPKGS_CONFIG=~/.config/nixpkgs/config-sm89.nix nix build --impure .#cudaPackages.saxpy
```

where `~/.config/nixpkgs/config-sm89.nix` is

```nix
{
  allowUnfree = true;
  cudaCapabilities = [ "8.9" ];
  cudaSupport = true;
}
```

ConnorBaker committed:
config: add and document {allow,block}listedLicenses

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
blender: OptiX is under lib.licenses.nvidiaCudaRedist

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
python3Packages.warp-lang: correct mathdx license

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

cpcloud commented:
I can build this now! Passing `--impure` did the trick.

cpcloud commented:
Nice, that works!

ConnorBaker committed:
openmpi: fix CUDA paths in configureFlags

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
_cuda.lib.{_mkMetaBadPlatforms,_mkMetaBroken}: use builtins.traceVerbose

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cudatoolkit: use all outputs but static from constituents

Fixes a bug where include and other outputs were not made available.

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker commented:
@cpcloud good catch -- the `cudatoolkit.nix` expression wasn't getting the `include` output, so headers largely weren't included in the output. I should have fixed it.

ConnorBaker committed:
python3Packages.cupy: fixup

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
suitesparse: correct CUDA_PATH

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
actiona: consume opencv CXX components when available

We do not default to `.dev` or any other output because `make-derivation.nix` needs to be able to select the `.dev` output,
and specifying an output prevents it from doing so.

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cudnn: 9.8.0 -> 9.13.0

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.libcusparse_lt: 0.7.1 -> 0.8.1

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.libcutensor: 2.2.0 -> 2.3.1

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

SomeoneSerge commented on pkgs/top-level/config.nix:
```diff
@@ -311,6 +357,8 @@ in
       mapAttrsToList (k: v: "undeclared Nixpkgs option set: config.${k}") config._undeclared or { }
     );
+
+    allowlistedLicenses = lib.mkIf config.cudaSupport cudaLicenses;
```
Sorry, I think we should never implicitly override or auto-complete user's consent choices, with the obvious exception of the "free" software licenses. In fact, I'm convinced that `allow*` configuration options are the perfect example of where we absolutely do need dynamic scopes: it should only be possible to set these options once at the top-level, and never from the imported code. Importantly, if we ever call `nixpkgsFun` with extended `allow*` definitions, we're no longer able to prove that `pkgs` does not depend on any of `pkgs{Unfree,Insecure,Broken}`.

"Implicitly overriding user's consents" is what an expression `pkgs.pkgsCuda.blender` would do with this change.

By introducing `pkgsForCudaArch` https://github.com/NixOS/nixpkgs/pull/406568, you implicitly started implementing the "Builder" pattern for `pkgs/top-level/config`. Following this logic, and ignoring the scoping concerns, you could introduce an explicit `allowUnfree` (`allowNvidia`) step into the path: https://github.com/NixOS/nixpkgs/pull/435082. Instead of polluting the top-level scope, we could introduce a dedicated entry point: `pkgs.CONFIGURE.allowUnfree.TRUE.allowLicense.APPEND.nvidiaCudaRedist.cudaSupport.TRUE.allowInsecurePnames.EMPTY.RESULT.blender`.

That said, I'd like to reiterate again, implementing a "`config` Builder" on the Nixpkgs side does not really solve the Flakes UX problem. For example, for the snippet above you'd have to not only obtain `attrNames lib.licenses`, but also obtain a "superset of all pnames in Nixpkgs", which may not be even computable.

To be completely frank, though, I do not believe that it's for Nixpkgs to be solving Flakes problems in the first place.

ConnorBaker replied:
> Sorry, I think we should never implicitly override or auto-complete user's consent choices, with the obvious exception of the "free" software licenses.

It would be explicit given the licenses accepted are documented in the description of the `cudaSupport` option.

> ... it should only be possible to set these options once at the top-level, and never from the imported code.

That is the case with Nixpkgs variants, as only they have access to `nixpkgsFun`. They exist purely as convenience attributes and avoid the need to maintain a bunch of different Nixpkgs config expressions (which can only be used one at a time and require impurity to do so).

> Importantly, if we ever call `nixpkgsFun` with extended `allow*` definitions, we're no longer able to prove that `pkgs` does not depend on any of `pkgs{Unfree,Insecure,Broken}`.

It is still possible to prove this, but it would require more work (an expression or script which runs over the closure given `meta` has all the data necessary to determine that) given evaluation won’t fail with an error. I agree that is concerning and problematic.

> "Implicitly overriding user's consents" is what an expression `pkgs.pkgsCuda.blender` would do with this change.

I think it was a mistake to add `pkgsCuda` and `pkgsForCudaArch` as Nixpkgs variants as they are -- the packages which they should enable evaluating and building (those which require CUDA) don't evaluate or build without providing external configuration (e.g., a `config` with `allowUnfree`, `allowUnfreePredicate`, or `allowlistedLicenses`).

I understand that I'm not representative of most users of the Nixpkgs CUDA packaging. I do, however, think most users are surprised that enabling `cudaSupport` isn't enough to evaluate or build anything relying on CUDA.

`pkgs.pkgsCuda.blender` would evaluate and build with this change.

> Following this logic, and ignoring the scoping concerns, you could introduce an explicit `allowUnfree` (`allowNvidia`) step into the path: https://github.com/NixOS/nixpkgs/pull/435082.

The discussion in that PR is interesting, thank you for the reference. At least as implemented currently using `nixpkgsFun`, chaining the builder pattern creates $O(n)$ instantiations of Nixpkgs, but I suppose that'd be re-architected if pursuing such an interface.

> To be completely frank, though, I do not believe that it's for Nixpkgs to be solving Flakes problems in the first place.

I agree; I think it's unfortunate that the UX is abysmal for anything which requires more than the default configuration.

I also think that we should be aiming to make the majority of use-cases just work; I think that if a user wants to build a package with CUDA, there should be an easy way to make that happen.

I don't think it's easy to create or maintain different `config` expressions, or use `--impure` and lose out on evaluation caching, or become familiar with every NVIDIA license required by the closure of the package they care about.

I do think this PR would make it easy for most people to use `pkgsCuda`.

adrian-gierakowski replied:
> It would be explicit given the licenses accepted are documented in the description of the cudaSupport option.

Maybe option should be named unfreeCudaSupport to make it obvious without having to read the docs

SomeoneSerge commented on pkgs/top-level/config.nix:
```diff
@@ -195,8 +195,44 @@ let
       '';
     };
 
+    allowlistedLicenses = mkOption {
+      description = ''
+        Allow licenses that are specifically acceptable. `allowlistedLicenses` only applies to unfree licenses unless
+        `allowUnfree` is enabled. It is not a generic allowlist for all types of licenses.
+      '';
+      default = [ ];
+      type = types.listOf (types.attrsOf types.anything);
+      example = literalExpression ''
+        with lib.licenses; [
+          amd
+          wtfpl
+        ]
+      '';
+    };
```
Rhetorical question: how does one access `lib.licenses` on the first `import nixpkgs` instantiation?

ConnorBaker replied:
I know you said rhetorical... but I only recently found out that `config` can also be a function which accepts an attribute set containing only `pkgs`:

https://github.com/NixOS/nixpkgs/blob/31c1c19e4828ef391f8d80310031de51306b26de/pkgs/top-level/default.nix#L114-L117

So... I guess you *can* use `pkgs` (and subsequently, `lib`) in `config`.

SomeoneSerge commented on pkgs/by-name/ac/actiona/package.nix:
```diff
@@ -44,7 +44,7 @@ stdenv.mkDerivation (finalAttrs: {
     bluez
     libnotify
-    opencv
+    opencv.cxxdev
```
Side-note: `lib.getFirstOutput ["cxxdev" "dev"] opencv`

ConnorBaker committed:
slurm: fixup usage of NVML

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
_cuda.lib.licenses: init

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages: remove module system evaluation

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages: remove generic-builder

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.buildRedist: init

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages: build redists from manifests

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cublasmp: init at 0.6.0

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.nvpl: init at 25.5

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.npp_plus: init at 0.10.0

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cudss: init at 0.6.0

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cuquantum: init at 25.09.0

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cusolvermp: init at 0.7.0

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.nvcomp: init at 5.0.0.6

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.libnvjpeg_2k: init at 0.9.0

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.libnvtiff: init at 0.5.1

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.tests.cmake: init

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.tests.cudnn-frontend: init

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.tests.onnx-tensorrt: init

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.tests.redists-unpacked: init

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.tests.redists-installed: init

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cudnn_samples: init

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cutlass: init at 3.9.2

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages_13: init at 13.0.2

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
python3Packages.torch: correct package name cusparselt -> libcusparse_lt

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
python3Packages.cupy: add libcusparse_lt now that it is available

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
_cuda.lib.allowUnfreeCudaPredicate: derive from _cuda.lib.licenses

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker commented:
More commit history cleanup; dropped the automatic addition of CUDA licenses to `allowlistedLicenses` when `cudaSupport` was enabled.

ConnorBaker committed:
tests.cuda: fixup

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker commented:
Something weird was going on with `clang*Stdenv` being unavailable in the `pkgs` attribute provided to `backendStdenv.nix` on Darwin (causing eval in CI to fail). Updated the way `backendStdenv.nix` selects a `stdenv` and a fallback in case we're on an unsupported platform (e.g., Darwin).

ConnorBaker commented:
Touched up the CUDA capabilities and rebased on master/fixed a merge conflict in `onnxruntime`.

Still haven't had a chance to rework the capability handling to support Thor. I'm fairly certain I'm going to need to rewrite `_cuda.lib.getRedistSystem` so it takes a CUDA version, list of CUDA capabilities, and Nix system instead of just a boolean for presence of a Jetson capability and Nix system.

ConnorBaker committed:
cuda-packages.nix: re-use backendStdenv.hasJetsonCudaCapability

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
_cuda.lib.getRedistSystem: refactor for CUDA 13 support

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.backendStdenv: clean up assertions for CUDA 13

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cuda_cudart: cuda_compat can be null or unavailable

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.autoAddCudaCompatRunpath: make hook a no-op when cuda_compat is unavailable

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.nccl: allow building on Thor and later

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker commented:
Okay, I started re-implementing `_cuda.lib.getRedistSystem`. I'm gonna need to put it in the changelog or something because the API did change.

General concerns I have now:

- document the difference (there isn't really one) between `backendStdenv.hasJetsonCudaCapability` and `flags.isJetsonBuild`
- CuPy `./pkgs/top-level/python-packages.nix` doing gross stuff to condition on cuDNN (maybe just unsupported by Thor or new releases until they finally drop support for cuDNN?)
- TensorRT `libcudla`
- `nsight_compute` and `nsight_systems` on Thor
- Usage of NCCL should be conditioned on `nccl.meta.available` and not `isJetsonBuild`
- `libcudla` -- probably shouldn't be enabled for all Jetsons as IIRC Thor doesn't have the DLA
- make sure changes to things didn't break existing builds (e.g., PyTorch et. al. still build and work with pre-13.0)
- fix up the documentation for all the various changes in this PR

ConnorBaker committed:
cudaPackages.tests.redists-{unpacked,installed}: include license metadata

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
xgboost: properly condition ncclSupport

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
python3Packages.cupy: fixup shouldUsePkg to check for broken-ness

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.libcudla: only available on Xavier and Orin

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.tensorrt: properly condition inclusion of libcudla

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker commented on pkgs/development/cuda-modules/_cuda/default.nix:
Add comment about manifests.

ConnorBaker commented on pkgs/development/cuda-modules/buildRedist/default.nix:
```diff
@@ -0,0 +1,484 @@
+        else
+          intersectLists outputs finalAttrs.passthru.expectedOutputs;
+
+      # Enforcing a size limitation on the out output lets us detect if we forgot to
+      # move something into the bin or lib output.
+      # This should be overridden by packages where we must keep more in `out`.
+      # TODO(@connorbaker): This doesn't seem to be working, or at least, isn't triggered when using `--rebuild`.
+      outputChecks.out.maxSize = 128 * 1024; # 128 KiB
```
Remove this and open an issue.

ConnorBaker commented on pkgs/development/cuda-modules/packages/autoAddCudaCompatRunpath/package.nix:
Should this change be flowed up and documented?

ConnorBaker commented on pkgs/development/cuda-modules/packages/collectx_bringup.nix:
Remove and document this one

ConnorBaker commented on pkgs/development/cuda-modules/packages/cuda_cudart.nix:
```diff
@@ -0,0 +1,99 @@
+    # TODO(@connorbaker): Check that the dependency offset for this is correct.
+    ++ [ (lib.getOutput "include" cuda_cccl) ]
+    ++
+      lib.optionals
+        (backendStdenv.hasJetsonCudaCapability && cuda_compat != null && cuda_compat.meta.available)
```
Is it not enough to just check for non-null and available?

ConnorBaker commented on pkgs/development/cuda-modules/packages/cuda_sandbox_dev.nix:
Remove and make a note

ConnorBaker commented on pkgs/development/cuda-modules/packages/cutlass.nix:
```diff
@@ -0,0 +1,232 @@
+    ]
+  );
+
+  postPatch =
```
Is it worth it to just vendor patches?

ConnorBaker commented on pkgs/development/cuda-modules/packages/driver_assistant.nix:
Remove and document this one

ConnorBaker commented on pkgs/development/cuda-modules/packages/fabricmanager.nix:
```diff
@@ -0,0 +1,19 @@
+    "lib"
+  ];
+
+  allowFHSReferences = true;
```
What's keeping them around?

ConnorBaker commented on pkgs/development/cuda-modules/packages/imex.nix:
```diff
@@ -0,0 +1,16 @@
+
+  outputs = [ "out" ];
+
+  allowFHSReferences = true;
```
What's keeping them around?

ConnorBaker commented on pkgs/development/cuda-modules/buildRedist/default.nix:
```diff
@@ -0,0 +1,484 @@
+            [ licenses.nvidiaCudaRedist ]
+          else
+            [ licenses.nvidiaCuda ];
+        teams = meta.teams or [ ] ++ [ teams.cuda ];
```
Make sure teams aren't specified in the redist packages with teams.cuda so we don't end up with double entries for the team.

ConnorBaker commented on pkgs/development/cuda-modules/packages/mft.nix:
Should this be removed?

ConnorBaker commented on pkgs/development/cuda-modules/packages/mft_oem.nix:
Should this be removed?

ConnorBaker commented on pkgs/development/cuda-modules/packages/nccl-tests.nix:
```diff
@@ -34,49 +40,74 @@ backendStdenv.mkDerivation (finalAttrs: {
   };
```
TODO: All FODs in `cudaPackages` should provide `name`, which must depend on version or something else which forces invalidation.

Also, switch to use of `tag`.

ConnorBaker commented on pkgs/by-name/bl/blender/package.nix:
```diff
@@ -436,9 +436,7 @@ stdenv'.mkDerivation (finalAttrs: {
-    license =
-      with lib.licenses;
-      [ gpl2Plus ] ++ lib.optional cudaSupport (unfree // { shortName = "NVidia OptiX EULA"; });
+    license = with lib.licenses; [ gpl2Plus ] ++ lib.optional cudaSupport nvidiaCudaRedist;
```
Double check that Optix is licensed as a redistributable.

ConnorBaker commented on pkgs/by-name/ac/actiona/package.nix:
```diff
@@ -44,7 +44,7 @@ stdenv.mkDerivation (finalAttrs: {
     bluez
     libnotify
-    opencv
+    (opencv.cxxdev or opencv)
```
I guess that also raises the issue of whether the `lib.getOutput` family of functions should be used within dependency arrays because they will select the `.out` output specifically as a fallback, when likely they should just return the derivation as is if the output doesn't exist.

SomeoneSerge replied:
Wait, on a closer look, this should be `getFirstOutput [ "cxxdev" "dev" ]`

ConnorBaker replied:
Remember mkDerivation tries to get the dev output of all dependency arrays, no need to do it manually

ConnorBaker commented on pkgs/development/cuda-modules/packages/nsight_compute.nix:
```diff
@@ -0,0 +1,105 @@
+let
+  qtwayland = lib.getLib qt6.qtwayland;
+  inherit (qt6) wrapQtAppsHook qtwebview;
+  archDir =
```
Still valid for thor?

ConnorBaker commented on pkgs/development/cuda-modules/packages/nsight_compute.nix:
```diff
@@ -0,0 +1,105 @@
+    rm -rf host/${archDir}/Mesa/
+  '';
+
+  postInstall = ''
```
add nixLog for these

ConnorBaker commented on pkgs/development/cuda-modules/packages/nsight_compute.nix:
```diff
@@ -0,0 +1,105 @@
+    moveToOutput 'target/${archDir}' "''${!outputBin}/bin"
+    wrapQtApp "''${!outputBin}/bin/host/${archDir}/ncu-ui.bin"
+  ''
+  # NOTE(@connorbaker): No idea what this platform is or how to patchelf for it.
+  + lib.optionalString (flags.isJetsonBuild && cudaOlder "12.9") ''
```
Add note about how 12.8 gets both applied

ConnorBaker commented on pkgs/development/cuda-modules/packages/nsight_systems.nix:
```diff
@@ -0,0 +1,147 @@
+    gst_all_1.gst-plugins-bad
+  ];
+
+  postInstall = ''
```
nixLog

ConnorBaker commented on pkgs/development/cuda-modules/packages/nvidia_driver.nix:
Remove this?

ConnorBaker commented on pkgs/development/cuda-modules/default.nix:
```diff
@@ -0,0 +1,152 @@
+      #   stdenv = pkgs'.gccStdenv;
+      # };
+
+      # Must be constructed without `callPackage` to avoid replacing the `override` attribute with that of
+      # `callPackage`'s.
+      buildRedist = import ./buildRedist {
```
Since everywhere it's used we're call-packaging it, does it matter if the override is clobbered? I think not. Move it back into the packages directory.

But the packages directory contains only packages... perhaps instead I should move backendStdenv out of packages?

ConnorBaker commented on pkgs/development/python-modules/cupy/default.nix:
```diff
@@ -87,17 +81,17 @@ buildPythonPackage rec {
   ];
 
-  NVCC = "${lib.getExe cudaPackages.cuda_nvcc}"; # FIXME: splicing/buildPackages
+  # NVCC = "${lib.getExe cudaPackages.cuda_nvcc}"; # FIXME: splicing/buildPackages
```
Can this environment variable be removed entirely?

ConnorBaker commented on pkgs/test/cuda/default.nix:
Ugh

ConnorBaker commented on pkgs/top-level/config.nix:
Make sure that specifying these as config options doesn't change evaluation semantics. (Note to self: this was observed while tooling around in `nix-nixpkgs-review` trying to generate attribute paths of packages which depend on CUDA.)

SomeoneSerge commented:
Can you confirm again, is this `config` being a fixpoint? If so, mayve add two three words to the comment above the snippet

On October 3, 2025 6:59:40 PM UTC, Connor Baker ***@***.***> wrote:
>@ConnorBaker commented on this pull request.
>
>
>
>On pkgs/top-level/config.nix:
>
>Make sure that specifying these as config options doesn't change evaluation semantics. (Note to self: this was observed while tooling around in `nix-nixpkgs-review` trying to generate attribute paths of packages which depend on CUDA.)
>

ConnorBaker committed:
_cuda.lib.allowUnfreeCudaPredicate: handle license not having shortName

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
colmap: only Linux is supported with CUDA

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
doc: update CUDA section

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages: remove some packages expressions which aren't ready

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.buildRedist: condition inclusion of autoAddCudaCompatRupath on its enablement

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cudnn-frontend: clean up expression

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cuda_nvcc: correct nvcc.profile patching for 13.0+

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cuda_nvcc: provide a single output

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cuda_cudart: provide a single output

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.autoAddCudaCompatRunpath: clean up enableHook

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cuda_cudart: don't symlink lib/stubs into lib

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cuda_nvml_dev: don't symlink lib/stubs into lib

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.gdrcopy: init at 2.5.1

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
ucc: fix linking against CUDA stubs

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
ucx: fix linking against CUDA stubs

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
python3Packages.gpuctypes: fix linking against CUDA stubs

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.libnvshmem: init at 3.4.5

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker commented:
Changes:

- rebased on `master`
- `cuda_cudart` and `cuda_nvml_dev` no longer symlink their stub into `lib` (in preparation for a hook to strip them after linking)
- introduced `gdrcopy`
- introduced `libnvshmem` (but not yet the Python library because what a pain that looks like)
  - building it from source instead of using the redist because source-based builds should always be preferred
  - looks like the python package, `nvshmem4py`, has dependencies on some stuff we've not yet packaged

Still need to update the `libcublasmp` package to make use of `libnvshmem`.

ConnorBaker committed:
gpu-burn: tidy up after CUDA changes

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.libcufile: fixup autoPatchelfHook for cuda_cudart stubs location

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
python3Packages.cupy: fix cuda_cudart stub linking

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
python3Packages.cupy: build without libcusparse_lt for now

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
koboldcpp: remove hardcoded link against CUDA driver

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.libcublasmp: use newly packaged libnvshmem

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.libcudla: buildRedist expects top-level platformAssertions

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker commented:
New changes:

- various other fixes as a result of stubs not being directly available within `lib`
- fixed cupy build by removing support for libcusparse_lt since cupy requires a very old version
- fixed the derivation for libcublasmp so it should build now that we've got libnvshmem

ConnorBaker committed:
cudaPackages.cuda_nvprof: fix cuda_cudart stub linking

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cuquantum: fix libnvidia-ml.so stub linking

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.nsight_systems: fix stub linking

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
python3Packages.tensorrt: fixup

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker commented:
Merge conflicts from the tree-wide `qt6.full` removal (#454156), I assume. I'll fix those later today.

ConnorBaker commented:
Rebased on `master`, removed the files mistakenly added back to `_cuda/fixups` by rebasing on the qt6 removal PR.

ConnorBaker committed:
cudaPackages.buildRedist: unpacking sources shouldn't depend on backendStdenv

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.tests.redists-installed: some packages may have dependents which are unavailable

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.autoAddCudaCompatRunpath: format with shfmt and lint with ShellCheck

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.markForCudatoolkitRootHook: format with shfmt and lint with ShellCheck

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.setupCudaHook: format with shfmt and lint with ShellCheck

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.cuda_cccl: restrict to a single output to fix libnvshmem build

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
_cuda.lib.selectManifests: init

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
pkgs/top-level/cuda-packages.nix: tidy using _cuda.lib.selectManifests

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
_cuda.db.bootstrap.cudaCapabilityToInfo: add 3.5, 3.7, and 7.2, and fix Orin minCudaMajorMinorVersion

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
_cuda.db.bootstrap.nvccCompatibilities: add back 11.0-12.5

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages: remove cudaFlags alias

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages: remove cudaVersion alias

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages: remove cudatoolkit-legacy-runfile alias

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages: remove cusparselt alias

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.tests.redists-installed: make broken assertion more descriptive

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages.nsight_compute: patch shebangs in scripts

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages: prefer autoPatchelfIgnoreMissingDeps to extraAutoPatchelfLibs

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

ConnorBaker committed:
cudaPackages: allow expressions to be used with older versions of CUDA

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

daniel-fahey commented on pkgs/development/cuda-modules/_cuda/manifests/tensorrt/helper.sh:
```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env bash
+
+# Requires nix and jq
```
```suggestion
#!/usr/bin/env nix-shell
#! nix-shell -i bash -p nix jq
```

daniel-fahey reviewed (commented):
Just a little nitpick suggestion

ConnorBaker committed:
cuda-modules: update TensorRT update helper

Signed-off-by: Connor Baker <ConnorBaker01@gmail.com>

emilazy commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -64,8 +64,8 @@ let
-  allowlist = config.allowlistedLicenses or config.whitelistedLicenses or [ ];
-  blocklist = config.blocklistedLicenses or config.blacklistedLicenses or [ ];
+  allowlist = config.allowlistedLicenses;
+  blocklist = config.blocklistedLicenses;
```
This is a breaking change, right? Is there any error message for someone who previously set `config.blacklistedLicenses`? I don’t mind removing it with a throw, but it seems like this will silently start permitting licences a user was specifically trying to opt out of.

To be honest I am not sure why release managers were not pinged in advance for a 99‐commit rework PR with many breaking changes merged this late after the freeze…

SomeoneSerge replied:
> reintroduce it after branch off

With an `mkRenamedOption` for `{black,white}listedLicenses`

ConnorBaker replied:
In hindsight, I should have used `mkRenamedOption`. I did not do so because they are undocumented attributes.

SomeoneSerge commented:
@ConnorBaker, I haven't been making it easier for you lately either, and didn't keep myself from being annoying in the few reviews that I sent. I can't quite put my finger at why, but I find just not being toxic is already hard enough.

> fundamental issues with the way several things are architected in Nixpkgs

We know, we all do...

> root problem

If I had to name just one, I couldn't.

I hope you can soon care again

ConnorBaker commented:
The @NixOS/cuda-maintainers have decided to revert the changes surrounding `config`:

- #457338
- #457220

@emilazy @wolfgangwalther @jopejoe1 if you have any questions or concerns about the CUDA packaging, we’re more than happy to schedule time to hear them!
